clang/tools/driver/CachedDiagnostics.cpp

```diff
@@ -0,0 +1,787 @@
+//===- CachedDiagnostics.h - Serializing diagnostics for caching ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is a mechanism for caching diagnostics with full source location
+// fidelity. It serializes the subset of SourceManager state for the diagnostic
+// source locations (FileIDs, macro expansions, etc.) and replays diagnostics by
+// materializing their SourceLocations; that way the diagnostic consumers have
+// full access to the include stack, expanded macros, etc.
+//
+// This allows us to only cache the diagnostic information that is essential for
+// the compilation and ignore the various ways that diagnostics can be rendered.
+//
+//===----------------------------------------------------------------------===//
+
+#include "CachedDiagnostics.h"
+#include "clang/Basic/DiagnosticCAS.h"
+#include "clang/Basic/FileManager.h"
+#include "clang/Basic/SourceManager.h"
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Compression.h"
+#include "llvm/Support/EndianStream.h"
+#include "llvm/Support/StringSaver.h"
+#include "llvm/Support/YAMLTraits.h"
+
+using namespace clang;
+using namespace clang::cas;
+using namespace llvm;
+
+namespace {
+
+namespace cached_diagnostics {
+
+struct Location {
+  unsigned SLocIdx;
+  unsigned Offset;
+};
+
+struct Range {
+  Optional<Location> Begin;
+  Optional<Location> End;
+  bool IsTokenRange;
+};
+
+struct FixItHint {
+  Optional<Range> RemoveRange;
+  Optional<Range> InsertFromRange;
+  std::string CodeToInsert;
+  bool BeforePreviousInsertions;
+};
+
+struct SLocEntry {
+  struct FileInfo {
+    std::string Filename;
+    Optional<std::string> Buffer;
+    Optional<Location> IncludeLoc;
+  };
+
+  struct ExpansionInfo {
+    Optional<Location> SpellingLoc;
+    Optional<Location> ExpansionStartLoc;
+    Optional<Location> ExpansionEndLoc;
+    unsigned Length;
+    bool IsTokenRange;
+  };
+
+private:
+  union {
+    FileInfo File;
+    ExpansionInfo Expansion;
+  };
+  bool IsExpansion;
+
+public:
+  SLocEntry() : File(), IsExpansion(false) {}
+  SLocEntry(FileInfo FI) : File(std::move(FI)), IsExpansion(false) {}
+  SLocEntry(ExpansionInfo EI) : Expansion(std::move(EI)), IsExpansion(true) {}
+
+  SLocEntry(const SLocEntry &SE) {
+    IsExpansion = SE.IsExpansion;
+    if (IsExpansion)
+      Expansion = SE.Expansion;
+    else
+      File = SE.File;
+  }
+  SLocEntry(SLocEntry &&SE) {
+    IsExpansion = SE.IsExpansion;
+    if (IsExpansion)
+      Expansion = std::move(SE.Expansion);
+    else
+      File = std::move(SE.File);
+  }
```

Comment:
With the rewrite from `std::variant` to the `union` and the `bool`, I think a small uninitialized memory bug was introduced in these two constructors. Since they don't have initialization list, they do their work over garbage memory, and the `File =` (and maybe the `Expansion =`) might try to free/copy over garbage memory.

Changing them to `SLocEntry(const SLocEntry &SE) : SLocEntry() {` and `SLocEntry(SLocEntry &&SE) : SLocEntry() {` seems to avoid the problem.

Reply:
Good catch! Fixed here: https://github.com/apple/llvm-project/pull/5842